salt/transport/rabbitmq.py

```diff
@@ -0,0 +1,1243 @@
+
+    def _on_connection_error(self, connection, exception):
+        """
+        Invoked by pika when connection on connection error
+        :param connection:
+        :param exception:
+        :return:
+        """
+        log.error("Failed to connect", exc_info=True)
+
+    def _on_channel_open(self, channel):
+        """
+        Invoked by pika when channel is opened successfully
+        :param channel:
+        :return:
+        """
+        self._channel.exchange_declare(exchange=self._fanout_exchange_name, exchange_type=ExchangeType.fanout, callback=self.on_exchange_declare)
+
+    def on_exchange_declare(self, method):
+        """Invoked by pika when RabbitMQ has finished the Exchange.Declare RPC
+        command.
+        """
+        log.debug("Exchange declared: {}".format(self._fanout_exchange_name))
+        self._channel.queue_declare(queue=self.queue_name, callback=self._on_queue_declared)
+
+    def _on_queue_declared(self, method):
+        """
+        Invoked by pika when queue is declared successfully
+        :param method:
+        :return:
+        """
+        log.debug("Queue declared: {}".format(self._queue_name))
+
+        self._channel.queue_bind(
+            self.queue_name,
+            self._fanout_exchange_name,
+            routing_key=self.queue_name,
+            callback=self._on_queue_bind)
+
+    def _on_queue_bind(self, method):
+        """
+        Invoked by pika when queue bound successfully
+        :param method:
+        :return:
+        """
+
+        log.debug("Queue bound [{}]".format(self.queue_name))
+
+        def _callback_wrapper(channel, method, properties, payload):
+             if self._callback:
+                log.debug("Calling message callback")
+                return self._callback(payload)
+
+        self._channel.basic_consume(self.queue_name, _callback_wrapper, auto_ack=True)
+
+    def register_message_callback(self, callback):
+        """
+        Register a callback that receives message on a queue
+        :param callback:
+        :return:
+        """
+        self._callback = callback
+
+
+class AsyncRabbitMQReqChannel(salt.transport.client.ReqChannel):
+    """
+    Encapsulate sending routines to RabbitMQ.
+
+    RMQ Channels default to 'crypt=aes'
+    """
+
+    # This class is only a singleton per minion/master pair
+    # mapping of io_loop -> {key -> channel}
+    instance_map = weakref.WeakKeyDictionary()
+    async_methods = [
+        "crypted_transfer_decode_dictentry",
+        "_crypted_transfer",
+        "_do_transfer",
+        "_uncrypted_transfer",
+        "send",
+    ]
+    close_methods = [
+        "close",
+    ]
+
+    def __new__(cls, opts, **kwargs):
+        """
+        Only create one instance of channel per __key()
+        """
+
+        # do we have any mapping for this io_loop
+        io_loop = kwargs.get("io_loop")
+        if io_loop is None:
+            io_loop = tornado.ioloop.IOLoop.current()
```

Comment:
**suggestion** this could also follow the pattern.

```
io_loop = kwargs.get("io_loop") or tornado.ioloop.IOLoop.current()
```

*typical* dict.get fallback would have it as `kwargs.get("io_loop", fallback_value)` but unlike the `or`, or ternary statements, the arguments to the function are evaluated first, so if there's a cost associated with `.current()` then it will be paid every `__new__`.